### PR TITLE
Augment inlined call stack with module type/instance info.

### PR DESCRIPTION
**Pull request by kimishpatel**
Summary:
Motivation:
During inlinng we lose information about where the node/op is coming from in module
hierarchy. The SourceDebug info is useful in providing where in the
source code the definition of the node exist. One can follow the
callstack to find the relevant module hierarchy. However, this is less
accessible, particularly in the context of aggregating node level op
latencies to produce module level latencies. Facilitiating this
enables identifying which module might be expensive and can be
targeted either for optimization or redesign. This also aids network
architecture search approaches.
Approach:
This PR adds module type/instance information to inlined call stack
during inlining. During inlining where the subsequent CallMethods are
being replaced with their corresponding graph, we find the module
instance information corresponding to the CallMethod following the
GetAttr node that produces the module type input to CallMethod.
This approach has limitations in that such information will not be
avaiable to CallFunctions and hence we will only be able to annotate
type information and not instance information.
But the intuition is that, it should cover majority of the cases. And we
should be able to address this in another way.

Test Plan:
python test/test_jit.py TestScopeWithInlinedCallSTack

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #{issue number}

### CODE REVIEW

File: torch/csrc/jit/ir/scope.h [hunk 1]
```diff
@@ -81,29 +103,45 @@ struct InlinedCallStack;
...
+      Function* fn,
+      SourceRange source_range,
+      c10::optional<ModuleInstanceInfo> module_instance_info);
+
   // Return next element in the callstack list.
   c10::optional<InlinedCallStackPtr> callee() const;
 
   // Return callstack as a vector of [Function, SourceRange] pairs.
   std::vector<InlinedCallStackEntry> vec();
+  std::vector<InlinedCallStackWithModuleInfo> vec_with_module_info();
```
Comment (ZolotukhinM) [inline torch/csrc/jit/ir/scope.h:144]: Since what you added is an optional field, I think instead of adding a new method `vec_with_module_info` we could just change the existing `vec`. AFAIK, there are no users of this yet anyway (except a couple of tests), so it should not break anything.

File: torch/csrc/jit/passes/inliner.cpp [hunk 2]
```diff
@@ -10,6 +10,51 @@ namespace prim {
...
+    }
+    else {
+      std::string free_fn = "FreeFunction::" + std::get<0>(tup)->name();
+      sc = sc->push(Symbol::scope(free_fn));
+    }
+  }
+  node->setScope(sc);
+}
+
+void ReconstructScopeFromInlinedCallStackHelper(Block* block) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/inliner.cpp:42]: Do we really need this pass (why can't we just work directly with `vec_with_module_info` where we need the full attribute path)? If yes, then it should be in a separate PR.

**Review by ZolotukhinM — APPROVED** [review]
Overall looks good! Please address the lint/clang-tidy complaints and also find a couple of remarks from me inline. You might want to ping @zdevito to get his opinion on this too :)

File: torch/csrc/jit/passes/inliner.cpp (hunk 2 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/inliner.cpp:42]: I mostly followed your pattern from reconstruct scopes, to actually update scope in a separate pass. Additionally it also helps to make testing easier. I can just run FileCheck on the graph.
It is possible that just call `ReconstructScopeFromInlinedCallStackHelper` directly after inlining, so you don't need to call this pass separately.
With respect to directly working with `vec_with_module_info`: That is certainly possible, but testing might be more complicated. How did you test `InlinedCallStack` changes?

Comment (kimishpatel) [reply torch/csrc/jit/passes/inliner.cpp:42]: Also I can split this up in a separate PR, but I dont know a clean way to test the base PR then.

File: torch/csrc/jit/ir/scope.h (hunk 1 above)
Comment (kimishpatel) [reply torch/csrc/jit/ir/scope.h:144]: That sounds good.

File: torch/csrc/jit/passes/inliner.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/inliner.cpp:42]: >  How did you test InlinedCallStack changes?

Here are the tests for `InlinedCallStack`: https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_misc.cpp#L1711-L1785
Probably, we could do a similar testing for this new functionality as well. I think it also would be a more rigorous testing as we wouldn't have a "man-in-the-middle" pass in the testing workflow.

I think I'm a bit reluctant with regard of this pass also because in a long term we were planning on getting rid of scopes entirely and replacing them with the InlinedCallStack, while this pass adds a new user of scopes. I don't think anyone is actively working in that direction, but it feels like a step in a slightly wrong direction.

Comment (kimishpatel) [reply torch/csrc/jit/passes/inliner.cpp:42]: I am ok with this. We can reconstruct the same thing with a helper function elsewhere during bytecode export. I will make the changes.

**Review by zdevito — COMMENTED** [review]
I am wary of this approach because introduces something very specific to 'Modules' into IR.
But IR is a level below this, where we reason about functions and objects. There is not requirement that there is a strict module hierarchy or that code has to run in a module (in fact torchscript functions run independently).

This would be equivalent to C++ or Python's debugger telling you which object something executed inside of instead of letting users see the stack trace. If users want to understand the performance of their models, it would be more clear to present. that performance in a way similar to perftools and other profilers which charge time to different parts of the stack trace. This is clearer than the module hierarchy because it can represent other abstractions as well, such as function calls or methods.

Things that aren't stack traces such as the old Scope object have been a huge headache for our team
because we do not know how they are getting used and because of that, when we change the IR we don't know how to fix this information. Adding additional debug context for modules will make this worse.

**Comment by kimishpatel** [comment]
> I am wary of this approach because introduces something very specific to 'Modules' into IR.
> But IR is a level below this, where we reason about functions and objects. There is not requirement that there is a strict module hierarchy or that code has to run in a module (in fact torchscript functions run independently).
> 
> This would be equivalent to C++ or Python's debugger telling you which object something executed inside of instead of letting users see the stack trace. If users want to understand the performance of their models, it would be more clear to present. that performance in a way similar to perftools and other profilers which charge time to different parts of the stack trace. This is clearer than the module hierarchy because it can represent other abstractions as well, such as function calls or methods.

In principle I agree with you. Most of the time perf debugging, for me, goes like that. However, some of our customers are interested in knowing things like, which modules are more expensive. For example, should I use transformer, or if so what might be the right size for it. And in those circumstances, I expect that they may do some kind of architecture search using module level information. If we expect model engineers to build automation tools to enable such architecture search, this feature might be important.
Also one can approach the same using traditional perf debug. Figure out the bottlenecks using stack trace and adjust your model structure accordingly. But building automation based on such an approach might be harder.
> 
> Things that aren't stack traces such as the old Scope object have been a huge headache for our team
> because we do not know how they are getting used and because of that, when we change the IR we don't know how to fix this information. Adding additional debug context for modules will make this worse.

I would pose it such. If you think that module level information is useful to have, for which I would take some liberty to presume yes, then how do we get to it. I am open to other ideas.
I guess we can also parse source tree/code from stack trace to build the same thing, but I dont know if that is the right way to go.
If you have time, I would love your input on that.
